### ISSUE TITLE
Crash when the orientation changes

### ISSUE THREAD

**Issue by sbaro** [michaelbarlow7/dungeon-crawl-android#55]
DCSS crashes when the phone or tablet orientation changes. Locking the orientation helps a bit, but not completely because when another application comes atop (for exemple due to an incoming phone call) the game crashes even with locked orientation. Changing the display size with the volume buttons also causes a crash.

The bug can be reproduced systematically. It happears on different hardwares with different android versions : for example on a LG Nexus 5 with Android 5.1.

**Comment by kribjo** [comment]
No such problems here on Nexus 5, Android 4.4.4, but I've noticed the same on Nexus 7 (2013) with Android 5.0.2. Both with version 0.16.1-android.
I would say this happens any time the app loses focus - opening and closing the notification drawer does not cause the problem but opening and closing the multitasking menu does.
There is no error message. aLogCat gives me nothing. Could DCSS itself be logging something somewhere?
ART related? Memory issue - standard Android out of memory behaviour?

**Comment by michaelbarlow7** [comment]
Thanks for the report guys, I've actually discovered and fixed the problem (in the gitorious submodule, so it won't show up in this repo).

I think the problem was ART related. I'm not 100% sure on the details, but basically each JNIEnv object you use is correlated with the thread in which its created, and only code called in its corresponding thread can use its relevant JNIEnv object. In the Game Activity's onResume(), I was calling the native code to send the current terminal screen back to the java code. This was being called via the UI thread, but the JNIEnv object stored belong to the game's thread, so it caused a crash. The method it called passed in a JNIEnv object though, so I just changed the method to use that instead.

This fix will be in the next build, currently working on fixing a few other tricky issues.
